File: frontend/src/constants/api-error-code.ts
/**
 * Đồng bộ với `com.dev.backend.resp.ApiErrorCode`.
 * Backend cũng có thể gửi `error` là tên HttpStatus (vd. BAD_REQUEST) khi AppException không set error.
 */
export const ApiErrorCode = {
  VALIDATION_ERROR: "VALIDATION_ERROR",
  TYPE_MISMATCH: "TYPE_MISMATCH",
  ACCESS_DENIED: "ACCESS_DENIED",
  INTERNAL_SERVER_ERROR: "INTERNAL_SERVER_ERROR",
} as const;

export type ApiErrorCode = (typeof ApiErrorCode)[keyof typeof ApiErrorCode];

export function isApiErrorCode(value: string): value is ApiErrorCode {
  return (Object.values(ApiErrorCode) as string[]).includes(value);
}

File: backend/src/main/java/com/dev/backend/resp/ApiErrorCode.java
package com.dev.backend.resp;

public final class ApiErrorCode {

    private ApiErrorCode() {}

    public static final String VALIDATION_ERROR = "VALIDATION_ERROR";
    public static final String TYPE_MISMATCH = "TYPE_MISMATCH";
    public static final String ACCESS_DENIED = "ACCESS_DENIED";
    public static final String UNAUTHORIZED = "UNAUTHORIZED";
    public static final String INTERNAL_SERVER_ERROR = "INTERNAL_SERVER_ERROR";
}
